pkg/server/container_execsync.go

```diff
@@ -48,92 +43,68 @@ func (c *criContainerdService) ExecSync(ctx context.Context, r *runtime.ExecSync
 	// Get container from our container store.
 	cntr, err := c.containerStore.Get(r.GetContainerId())
 	if err != nil {
-		return nil, fmt.Errorf("an error occurred when try to find container %q: %v", r.GetContainerId(), err)
+		return nil, fmt.Errorf("failed to find container in store: %v", err)
 	}
 	id := cntr.ID
 
 	state := cntr.Status.Get().State()
 	if state != runtime.ContainerState_CONTAINER_RUNNING {
-		return nil, fmt.Errorf("container %q is in %s state", id, criContainerStateToString(state))
+		return nil, fmt.Errorf("container is in %s state", criContainerStateToString(state))
 	}
 
-	// Get exec process spec.
-	container, err := c.containerService.Get(ctx, id)
+	// TODO(random-liu): Store container client in container store.
+	container, err := c.client.LoadContainer(ctx, id)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get container %q from containerd: %v", id, err)
+		return nil, fmt.Errorf("failed to load container: %v", err)
 	}
-	var spec runtimespec.Spec
-	if err := json.Unmarshal(container.Spec.Value, &spec); err != nil {
-		return nil, fmt.Errorf("failed to unmarshal container spec: %v", err)
+	spec, err := container.Spec()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get container spec: %v", err)
 	}
-	pspec := spec.Process
-	pspec.Args = r.GetCmd()
-	rawSpec, err := json.Marshal(pspec)
+	task, err := container.Task(ctx, nil)
 	if err != nil {
-		return nil, fmt.Errorf("failed to marshal oci process spec %+v: %v", pspec, err)
+		return nil, fmt.Errorf("failed to load task: %v", err)
 	}
 
-	// TODO(random-liu): Replace the following logic with containerd client and add unit test.
-	// Prepare streaming pipes.
-	execDir, err := ioutil.TempDir(getContainerRootDir(c.rootDir, id), "exec")
+	pspec := spec.Process
+	pspec.Args = r.GetCmd()
+
+	execID := generateID()
+	stdinBuf, stdoutBuf, stderrBuf := new(bytes.Buffer), new(bytes.Buffer), new(bytes.Buffer)
+	io := containerd.NewIOWithTerminal(stdinBuf, stdoutBuf, stderrBuf, pspec.Terminal)
+	process, err := task.Exec(ctx, execID, pspec, io)
 	if err != nil {
-		return nil, fmt.Errorf("failed to create exec streaming directory: %v", err)
+		return nil, fmt.Errorf("failed to create exec %q: %v", execID, err)
 	}
 	defer func() {
-		if err = c.os.RemoveAll(execDir); err != nil {
-			glog.Errorf("Failed to remove exec streaming directory %q: %v", execDir, err)
+		if _, err := process.Delete(ctx); err != nil {
+			glog.Errorf("Failed to delete exec process %q for container %q: %v", execID, id, err)
```

Comment:
still a sanity check won't hurt... you can add this should not happen if you like to the error message that is un-necessary :-)